### ISSUE TITLE
config passwd = yes   of no effect or I misunderstood its purpose?

### ISSUE THREAD

**Issue by yarikoptic** [apptainer/singularity#125]
built a fresh debian package (from the state of #121), installed, seems to work!  if I got it correct, default configuration has config passwd = yes and that should have created a user entry in the /etc/passwd for me, but I think it isn't happening:

``` shell
$> grep passwd /etc/singularity/singularity.conf 
# If /etc/passwd exists within the container, this will automatically append
config passwd = yes

$> grep -1 passwd /etc/singularity/singularity.conf 
# DEFAULT: yes
# If /etc/passwd exists within the container, this will automatically append
# an entry for the calling user
config passwd = yes

$> whoami                                          
yoh

$> singularity shell /tmp/container.img     
Singularity/container.img> whoami
whoami: cannot find name for user ID 47521
```

container was freshly generated using `sudo singularity bootstrap /tmp/container.img examples/debian.def
` . I was expecting for user yoh to be known inside the container

**Comment by gmkurtzer** [comment]
can you show me the bottom few lines of the /etc/passwd in your container and `grep yoh /etc/passwd`?

Also, can you try to run singularity with debugging enabled:

`singularity --debug shell /tmp/container.img`

and send me the output?

Thanks!

**Comment by yarikoptic** [comment]
well -- when I run it without sudo, permission is denied to access
/etc/passwd:

``` shell
$> singularity shell /tmp/neurodebian.img
Singularity/neurodebian.img> cat /etc/passwd
cat: /etc/passwd: Permission denied
Singularity/neurodebian.img> ls -ld /etc/passwd
-rw------- 1 0 0 1257 Jul  8 20:24 /etc/passwd
```

and even if I make it readable to everyone while under root, upon then non-root invocation -- it is not readable again:

``` shell
$> sudo singularity shell --writable /tmp/neurodebian.img                                                 
[sudo] password for yoh:  
Singularity/neurodebian.img> chmod a+r /etc/passwd
Singularity/neurodebian.img> ls -l /etc/passwd
-rw-r--r-- 1 root root 1197 Jul  8 17:01 /etc/passwd
Singularity/neurodebian.img> exit

$> singularity shell /tmp/neurodebian.img           
Singularity/neurodebian.img> tail /etc/passwd
tail: cannot open '/etc/passwd' for reading: Permission denied
Singularity/neurodebian.img> ls -ld /etc/passwd
-rw------- 1 0 0 1257 Jul  8 20:26 /etc/passwd
Singularity/neurodebian.img> exit
```

output of debug sent via email (does it have any privatish data? on a quick look didn't see)

**Comment by gmkurtzer** [comment]
Well, that is weird... What is your umask set to?

Long story short, when singularity is called as a non-root user,
/etc/passwd and /etc/group within a container are modified outside the
container and bind'ed back into the container to account for the user
calling the container. /etc/passwd|group are copied from the container to
the session directory, then lines are appended to those files referencing
the calling user, and then those files are bounded back into the container
ontop of the originals. Soo... What I think might be going wrong would have
to do with umask. I have to check the code, but it is possible I didn't
preserve the permissions of those files when I created the copies.

What is your umask set to? Can you open it up and see if that fixes it?

**Comment by yarikoptic** [comment]
On Fri, 08 Jul 2016, Gregory M. Kurtzer wrote:

>    Well, that is weird... What is your umask set to?

here we are talking! ;)   mine is the one which gives pain more often
than it helped I think --  077  ;)

>    Long story short, when singularity is called as a non-root user,
>    /etc/passwd and /etc/group within a container are modified outside the
>    container and bind'ed back into the container to account for the user
>    calling the container. /etc/passwd|group are copied from the container to
>    the session directory, then lines are appended to those files referencing
>    the calling user, and then those files are bounded back into the container
>    ontop of the originals. Soo... What I think might be going wrong would
>    have
>    to do with umask. I have to check the code, but it is possible I didn't
>    preserve the permissions of those files when I created the copies.
> 
>    What is your umask set to? Can you open it up and see if that fixes it?

opened up!  besides some empty line in the tail of /etc/passwd:

systemd-bus-proxy:x:103:106:systemd Bus Proxy,,,:/run/systemd:/bin/false

yoh:x:47521:47522:Yaroslav Halchenko,,,:/home/yoh:/bin/zsh

it all works -- whoami works and permissions are set sanely.

DO NOT TRUST USERS HAVING A SANE UMASK -- but preserve it inside I guess  ;)

## Cheers!

Yaroslav O. Halchenko
Center for Open Neuroscience     http://centerforopenneuroscience.org
Dartmouth College, 419 Moore Hall, Hinman Box 6207, Hanover, NH 03755
Phone: +1 (603) 646-9834                       Fax: +1 (603) 646-1419
WWW:   http://www.linkedin.com/in/yarik

**Comment by gmkurtzer** [comment]
Fixed in [master 2a7c530], thanks!
